homeassistant/components/turn_touch.py

```diff
@@ -0,0 +1,166 @@
+
+REQUIREMENTS = ['turntouch==0.4.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_DEBOUNCE = 'debounce'
+
+DOMAIN = 'turn_touch'
+DATA_KEY = 'DATA_{component}'.format(component=DOMAIN)
+
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_MAC): cv.string,
+    vol.Optional(CONF_DEBOUNCE, default=True): cv.boolean,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_DEVICES, default=[]): [DEVICE_SCHEMA],
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+EVENT_NAME = 'turntouch_press'
+EVENT_DATA_ADDRESS = 'remote_address'
+EVENT_DATA_ENTITY_ID = 'entity_id'
+EVENT_DATA_NAME = 'remote_name'
+EVENT_DATA_TYPE = 'press_type'
+
+RETRY_DELAY = .25
+
+TURN_TOUCH_PLATFORMS = ['sensor']
+
+
+def setup(hass, config):
+    """
+    Set up the Turn Touch platform.
+
+    Also, set up other components (eg. sensor) provided by this platform.
+    """
+    hass.data[DATA_KEY] = {'devices': {}}
+
+    for device_config in config[DOMAIN][CONF_DEVICES]:
+        turn_touch = TurnTouchRemote(hass, device_config)
+        hass.data[DATA_KEY]['devices'][turn_touch.address] = turn_touch
+
+    for platform in TURN_TOUCH_PLATFORMS:
+        load_platform(hass, platform, DOMAIN, {}, config)
+
+    return True
+
+
+class TurnTouchRemote:
+    """Representation of a Turn Touch remote.
+
+    This does not inherit from the Entity class because Entities are created
+    by platforms (e.g., sensor.turn_touch). But we must maintain one shared
+    connection to the device, so that happens here.
+    """
+
+    def __init__(self, hass, config, listen=True):
+        """Initialize the remote."""
+        # pylint: disable=import-error
+        import turntouch
+        self._hass = hass
+        self.address = config[CONF_MAC]
+        self._debounce = config[CONF_DEBOUNCE]
+        self._device = None
+
+        class Handler(turntouch.DefaultActionHandler):
+            """Callback handler for button presses."""
+
+            def __init__(self, handler_func):
+                self._handler_func = handler_func
+
+            def action_any(self, action: turntouch.Action = None):
+                """Handle all button presses."""
+                if not action.is_off:
+                    self._handler_func(action.name)
+
+        self._handler = Handler(self._button_press)
+        self._connect()
+        self.name = self.get_name()
+        self._battery = None  # may not be needed; save time & get it later.
+        threading.Thread(target=self._listen).start()
+
+    def get_name(self):
+        """Read the device name."""
+        # pylint: disable=import-error
+        import turntouch
+        delay = RETRY_DELAY
+        while True:
+            try:
+                self.name = self._device.name
+                return self.name
+            except turntouch.TurnTouchException:
+                _LOGGER.warning('Reading device name failed. Retrying...')
+            time.sleep(delay)
```

Comment:
Sleeping in functions executed in the home assistant tread pool is not allowed. `get_name` is called inside component setup, ie the thread pool. It's ok to sleep in a separate thread. Async sleep is also ok.